source/credentials_provider_imds.c

```diff
@@ -737,14 +740,29 @@ static imds_state_fn *s_query_state_machine[] = {
     [AWS_IMDS_QS_ROLE_CREDENTIALS_REQ] = s_imds_query_instance_role_credentials_req,
     [AWS_IMDS_QS_ROLE_CREDENTIALS_RESP] = s_imds_query_instance_role_credentials_response,
     [AWS_IMDS_QS_COMPLETE] = s_imds_query_complete,
+    [AWS_IMDS_QS_QUERY_NEVER_CLEARED_STACK] = s_imds_query_error,
     [AWS_IMDS_QS_UNRECOVERABLE_ERROR] = s_imds_query_error,
 };
 
+static inline void s_imds_state_machine_roll_back_to_request_state(
+    struct aws_credentials_provider_imds_user_data *user_data) {
+    AWS_FATAL_ASSERT(
+        user_data->query_state < AWS_IMDS_QS_COMPLETE &&
+        "State machine can't be advanced, it's already in a terminal state.");
+    user_data->query_state -= 1;
+    /* request states are evenly numbered. */
+    AWS_FATAL_ASSERT(!(user_data->query_state & 0x01));
```

Comment:
If we're adding string context on the assert, might as well indicate something like "cannot rollback from request state"